bank/handlers.go

```diff
@@ -0,0 +1,380 @@
+package bank
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/gorilla/context"
+	"github.com/gorilla/sessions"
+	"golang.org/x/crypto/bcrypt"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"log"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+type UserRepo struct {
+	C                   []*mgo.Collection
+	Store               *sessions.CookieStore
+	renewSessionCounter chan bool
+}
+
+type User struct {
+	ID       bson.ObjectId `bson:"_id,omitempty"`
+	Username string        `bson:"username"`
+	Password string        `bson:"password"`
+	Ssn      int           `bson:"ssn"`
+	Name     string        `bson:"name"`
+	Email    string        `bson:"email"`
+	Age      int           `bson:"age"`
+	Hash     []byte        `bson:"hash"`
+}
+
+type Account struct {
+	ID              bson.ObjectId `bson:"_id,omitempty"`
+	Username        string        `bson:"username"`
+	Currency        string        `bson:"currency"`
+	Amount          float64       `bson:"amount"`
+	TransactionType string        `bson:"transactiontype"`
+}
+
+type Session struct {
+	ID       bson.ObjectId `bson:"_id,omitempty"`
+	Expires  time.Time     `bson:"expires"`
+	Username string        `bson:"username"`
+}
+
+func NewUserRepository(s mgo.Session, dbName string, colName ...string) *UserRepo {
+	repo := new(UserRepo)
+	for _, p := range colName {
+		c := s.DB(dbName).C(p)
+		repo.C = append(repo.C, c)
+	}
+	repo.renewSessionCounter = make(chan bool)
+
+	return repo
+}
+
+func (u *User) hashGenerator() error {
+	var err error
+	u.Hash, err = bcrypt.GenerateFromPassword([]byte(u.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (us *UserRepo) authenticate(u User) error {
+	result, err := us.GetUser(u)
+	if err != nil {
+		return err
+	}
+
+	if len(result) == 0 {
+		return errors.New("username not found")
+	}
+
+	// Comparing the password with the hash
+	if err = bcrypt.CompareHashAndPassword(result[0].Hash, []byte(u.Password)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (us *UserRepo) voidSessionAfter(u User, w http.ResponseWriter) {
+	select {
+	case <-time.After(time.Second * 300):
+		err := us.DeleteSession(u)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+		return
+	case renew := <-us.renewSessionCounter:
+		if renew == true {
+			us.voidSessionAfter(u, w)
+			return
+		} else {
+			return
+		}
+	}
+}
+
+func (us *UserRepo) TransactionsHandler(w http.ResponseWriter, r *http.Request) {
+	a := context.Get(r, "account")
+	s := context.Get(r, "session")
+
+	request := a.(Account)
+	gorillaSession := s.(*sessions.Session)
+
+	switch r.Method {
```

Comment:
gorilla mux could route HTTP methods separately which removes boiling checks for methods in the handlers. 

Here is an example how this is accomplished in main.go: 
```golang
// connect to db 
db, err := mongo.NewDatabase(config)
if err != nil {
  log.Fatalf("could not connect to database due: %v", err)
}

// zone repository is created from db
zoneRepo := persistent.NewZoneRepository(db)


apiRoot := mux.NewRouter().PathPrefix("/api/v1").Subrouter()
apiRouter.Handle("/zones", api.CreateZone(zoneRepo)).Methods("POST")
apiRouter.Handle("/zones", api.FetchAllZones(zoneRepo)).Methods("GET")
apiRouter.Handle("/zones/{id}", api.FetchZoneByID(zoneRepo)).Methods("GET")
apiRouter.Handle("/zones/{id}", api.UpdateZone(zoneRepo)).Methods("PUT")
```

More information could be taken from:
http://www.gorillatoolkit.org/pkg/mux